src/cascade/model/object_wrapper.py

```diff
@@ -0,0 +1,315 @@
+from collections import Iterable
+from contextlib import contextmanager
+from math import nan, isnan
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+
+from cascade.core import getLoggers
+from cascade.dismod.constants import DensityEnum, RateEnum
+from cascade.dismod.db.wrapper import DismodFile, get_engine
+from cascade.dismod.serialize import default_integrand_names, make_log_table
+from cascade.model.data_read_write import (
+    write_data, avgint_to_dataframe, read_avgint, read_data_residuals, read_simulation_data
+)
+from cascade.model.grid_read_write import (
+    read_var_table_as_id, read_vars, write_vars, read_prior_residuals, read_samples,
+    read_simulation_model
+)
+from cascade.model.model_writer import ModelWriter
+
+CODELOG, MATHLOG = getLoggers(__name__)
+
+
+class ObjectWrapper:
+    """
+    An I/O layer on top of the Dismod db file that presents Model objects.
+    It sets and gets models, vars, data, and residuals.
+    """
+    def __init__(self, locations, parent_location, filename):
+        assert isinstance(locations, pd.DataFrame)
+        assert isinstance(parent_location, int)
+        assert isinstance(filename, (Path, str))
+
+        self._filename = Path(filename)
+        self.dismod_file = None
+        self.parent_location = parent_location
+        self.location_func = None
+        self._locations_df = locations
+
+        # From covariate name to the x_<number> name that is used internally.
+        # The session knows this piece of information but not the covariate
+        # reference values. This is here because the columns of avgint and data
+        # need to be renamed before writing, and they aren't part of the model.
+        self._covariate_rename = dict()
+
+    @property
+    def model(self):
+        raise NotImplementedError("Cannot read a model.")
+
+    @model.setter
+    def model(self, new_model):
+        """When you write a model, it deletes the file."""
+        self.make_new_dismod_file(self._locations_df)
+        writer = ModelWriter(self, {})
+        new_model.write(writer)
+        writer.close()
```

Comment:
The business about making a new file... should flush every time you write anything. It's called isolation level for sqlite3. "exclusion" seems to be right. Take a look. (Greg again)